### ISSUE TITLE
Location based AR in Three.js (+ React) - camera configuration?

### ISSUE THREAD

**Issue by creimers** [jeromeetienne/AR.js#675]
I'm making this a feature request issue because it is for sure not a bug...

**Is your feature request related to a problem? Please describe.**

I have an existing React app that displays among other things a 120 deg static panorama (taken from five rgb cameras which don't move). The panorama is fed through object detection, and the bounding boxes of the detected objects are rendered as an overlay on the panorama. A geolocator algorithm furthermore provides rather really accurate geolocation estimates of the detected objects. This is where AR comes into play: I want to create an additional overlay with a geospatial representation of the detected objects --> location based AR.

As three.js is already used in another part of the app (and it has an excellent [React implementation](https://github.com/react-spring/react-three-fiber/)), I want to stick to it.

**Describe the solution you'd like**

Looking at [GeoAr.js](https://github.com/nicolocarpignoli/GeoAR.js/blob/master/src/gps-entity-place.js#L43), it has been relatively straightforward to implement the rendering of the location markers (ignoring the [pros and cons of the used algorithm](https://github.com/jeromeetienne/AR.js/issues/663#issuecomment-560033972) for now), and when looking at it "standalone", it looks _right_. When I put this on top of my panorama, however, the locations of the AR markers and the "real life" objects in the panorama don't match too well. My hypothesis is that this is due to the camera settings (fov, near, far, zoom, ...) I have chosen (my knowledge of computer graphics is rather poor).

So the **core of my request** really is: Can someone (@nicolocarpignoli ?) maybe hint at how to configure the camera (or the scene in general) in order to achieve a good visual match of the real life objects and the geolocation markers? 

**Additional context**

I cannot provide screenshots of the app unfortunately, but I'm offering to summarize and publish my approach (React + three.js + location based AR) in case I get this up and running, thus make my contribution to the community.

**Comment by evaristoc** [comment]
@creimes 
Current implementation of GeoAR.js works on a-frame, which is a framework based on three.js. The camera settings can be indeed configurated by tuning the parameters of the a-camera component (https://aframe.io/docs/0.9.0/primitives/a-camera.html).

For AR markers at ' the locations of the AR markers and the "real life" objects in the panorama' I believe you refer to the tooltip markers currently offered by the application?

For what I see, your main issue is the matching between your frame of reference and the one by the implementation. The current implementation of GeoAR.js estimates a ["distance" between coordinates based on current location]( https://github.com/nicolocarpignoli/GeoAR.js/blob/8e399867760570acfe991e20dd38589a6c8a66d8/src/gps-camera.js#L197). The distance is translated into the world frame of reference of the current implementation to determine position of the objects (x,z coordinates).

I am not sure about the precision of the location/distance metrics. For what I have read in other threads and references, the precision might be low and change along the time. There are suggestions it could have an error of between 5-20 meters. That could be a reason for your markers not falling in the right position in your panorama. Additionally, I guess you will find differences between devices.

IMO, the measure based on latitude/longitude wouldn't comply with a proper perspective calculation: it might be good for reads like "it is closer" or "it is far" but in one of my tests the marker grew so big when approaching it covering the full screen - and it was NOT in the exact location. Furthermore, see that the data might not be estimated for y (altitude), so it is not a correct measure for perspective.

I think the challenge would be how to correct the positioning giving additional "clues" to your program. Just on the fly, one positive thing is that you have the correct positional order of the objects and a bit of perspective. Besides you might get relevant data given it is a static panorama. For example: you need at least 2 measures from your visuals to estimate a correction of the perspective on monocular systems (see "Familiar Size" at https://en.wikipedia.org/wiki/Monocular_vision).

I wonder if it wouldn't be better to think your problem from the static panorama of yours, actually, instead of a geolocation: what about a pre-mapping of the the geocoordinates into your static panorama, so once you get the coordinates you can position the marker?

To be frank, I haven't tried any problem like yours: I am just suggesting possible startpoints. Maybe @nicolocarpignoli has a better conceived idea?

**Comment by nicolocarpignoli** [comment]
The problem here, if I got this right, it is not about precision of GPS or GeoAR.js 'standalone', that for the thread author is good enough. Problem is when an AR.js -based app (in this case, geolocation based) is wrapped on another page, possibly created with a JS Framework (e.g. React).

This is a known bug that occurs also with marker based AR.js. It's an issue related to canvas used by AR.js I guess.

I had never time to look into that. I can get some reference though: https://github.com/jeromeetienne/AR.js/issues/537

**Comment by nicolocarpignoli** [comment]
To test if my thesis is correct, you can try to implement a basic marker based app in  the geolocation app you already have and see if the problem is similarly happening.

**Comment by evaristoc** [comment]
@nicolocarpignoli @creimers yea... it can be that my interpretation of the problem is incorrect. I apologise if that is the case.

**Comment by creimers** [comment]
@nicolocarpignoli @evaristoc 

Thanks for your input. I'm sorry if my initial description was not as clear as intended. Let me try again.

Essentially, I want to do location based AR more or less just like AR.js provides it out of the box. I have, however, a number of constraints/requirements:

* I want to add location based AR to an existing React app.
* I want to use three.js instead of aframe, because three.js is already used in other parts of the app, and because it has an excellent React implementation (which aframe does not have).
* Instead of placing the location based AR markers on top of a mobile device's camera input, I want to overlay them onto a panorama obtained from an RGB camera.
* The position of the camera is known very precisely (x, y, z).
* The location of the AR targets I want to display is known very precisely.

Having said that, I do have a basic implementation up and running. I will create an online example to look at later.

While I can calculate the positions of the AR targets and put them in a scene just fine, I'm having trouble to achieve a good match between the generated AR targets and the "real live objects". My hypothesis is that this is due to camera/scene settings.

**Comment by denisstasyev** [comment]
@creimers You can check my repository with the results of my work and tips for embedding AR.js in React. You can find it [here](https://github.com/denisstasyev/AR-Test-2). Hope it will help you.

It contains an implementation of AR.js as a React Component for any projects. You can see this component [here](https://github.com/denisstasyev/AR-Test-2/blob/master/src/components/ViewerVR/ViewerVR.js).

**Comment by nicolocarpignoli** [comment]
@creimers location-based feature is actually only available for a-frame

**Comment by creimers** [comment]
@nicolocarpignoli I understand that the location-based feature is only available for a-frame in AR.js. 

But I should be able to completely recreate it using three.js, no?

**Comment by nicolocarpignoli** [comment]
Yes absolutely

**Comment by nicolocarpignoli** [comment]
If you do that please provide a PR

**Comment by creimers** [comment]
@nicolocarpignoli Sure sure. 

I have finally created a [sandbox](https://codesandbox.io/s/wonderful-kalam-msmji?fontsize=14&hidenavigation=1&module=%2Fsrc%2Fcomponents%2FARScene%2Findex.jsx&theme=dark) to illustrate what I'm struggling with (never mind the marker offset on the map). I can calculate the position of the objects no problem, but I don't manage to configure the scene/camera in a way that the actually optically match with what's in the static image. The field of view of the camera should be ~55 degrees (iPhone XR).

If you were to find the time to have a look and/or point me to the juicy bits of camera configuration in the a-frame implementation, it would be greatly appreciated. I promise I will publish my three.js based approach once I'm succeeding.

**Comment by evaristoc** [comment]
@creimers sorry but I still don't understand. What do you mean with:

**Comment by evaristoc** [comment]
@creimers 

Thanks for sharing the sandbox. Let me see if I am understanding your problem...

You wrote:
> React app that displays among other things a 120 deg static panorama (taken from five rgb cameras which don't move)
> when looking at it "standalone", it looks right. When I put this on top of my panorama, however, the locations of the AR markers and the "real life" objects in the panorama don't match too well.

My first guess is that the panorama view deforms the space? If so, it could be that the three.js/a-frame SCENE used to render the AR markers is probably NOT deformed as the panorama.

Assuming I am right, again haven't dealt with something similar, here my list of things you could start exploring:
1. there is an attribute of the camera object, "aspect" that could help you to tweak the aspect ratio of the three.js camera; my first guess is that it should closely correspond to the aspect ratio of your panorama.
2. bending your scene by rendering your objects over the arc of a sphere that is an appropriate reflection of your panorama view - I think this would help to render them facing to the right direction.
3. think about how to recalculate the coordinates so they fit into a panoramic (deformed) view.

In case it is the scene what it is causing an issue, I would suggest you to visit the three.js discourse forum with this problem? And I hope I have been of help.

**Comment by evaristoc** [comment]
@creimers I also found this: https://medium.com/@kumar.ahir/how-to-correctly-set-dimensions-of-an-image-in-a-frame-8c1e598416e1

**Comment by evaristoc** [comment]
I was playing with your values and I think I am getting your problem.

@creimers What for camera is that? Camera class, Ortographic or Perspective? Can you please check and probably change it to see if there is an improvement?

**Comment by creimers** [comment]
@evaristoc Thanks for your thoughts, ideas and suggestions. I will evaluate them and report.

Regarding the camera: I'm currently using a PerspectiveCamera. I tried the OrthographicCamera, too, but it seems to be of no use.

**Comment by evaristoc** [comment]
@creimers Nice. Just after turning off my computer I realised that indeed you are rendering the objects along a plane, but your panorama is curving the view.

I currently don't remember the name of that deviation but it is common phenomenon to all panoramic views. I have read something about it when investigating about VR scenes. If I manage to recall the reference I will share with you all.

I think that bending of the view will affect all proportions and distances. In particular to your project I noticed there is a deformation of the ground that is clearer on the sides of the image. You have to find a way to deform your three.js rendering plane so it reflects that bending.

I will follow the discourse post as I am interested in your problem.

**Comment by evaristoc** [comment]
@creimers you might want to try to correct your panoramic view instead? Have a look at https://github.com/jywarren/fisheyegl. The opposite is also valid (ie. distort your three.js scene so it fits your panorama image) but that appears to be a bit more complicated and it is a question for the three.js chat group.

**Comment by nicolocarpignoli** [comment]
we can continue on new AR.js repo: https://github.com/AR-js-org/AR.js/issues/31
